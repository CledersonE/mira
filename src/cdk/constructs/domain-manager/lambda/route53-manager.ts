import { Context, SNSEvent } from 'aws-lambda'
import sendCfnResponse, { LambdaEvent, ResponseData } from './send-cfn-response'
import { createRecord, deleteRecord, upsertRecord } from './route53-record-actions'

export const handler = async (event: SNSEvent, context: Context): Promise<string> => {
  console.log(`SNS event: ${JSON.stringify(event)}`)
  const lambdaEvent = JSON.parse(event.Records[0].Sns.Message) as LambdaEvent

  try {
    const hostedZone = process.env.HOSTED_ZONE || ''
    if (!hostedZone) throw new Error('Hosted Zone not set')
    const type = lambdaEvent.RequestType
    const source = lambdaEvent.ResourceProperties.Source
    const target = lambdaEvent.ResourceProperties.Target

    const responseData: ResponseData = {}
    switch (type) {
      case 'Create': {
        console.log(`Creating CNAME ${source} -> ${target} in hosted zone: ${hostedZone}`)
        await createRecord(hostedZone, source, target)
        break
      }
      case 'Update': {
        const oldsource = lambdaEvent.OldResourceProperties.Source
        const oldtarget = lambdaEvent.OldResourceProperties.Target
        console.log(`Deleting old CNAME ${oldsource} -> ${oldtarget} in hosted zone: ${hostedZone}`)
        await deleteRecord(hostedZone, oldsource, oldtarget)
        console.log(`Creating new CNAME ${source} -> ${target} in hosted zone: ${hostedZone}`)
        await upsertRecord(hostedZone, source, target)
        break
      }
      case 'Delete': {
        console.log(`Deleting CNAME ${source} -> ${target} in hosted zone: ${hostedZone}`)
        await deleteRecord(source, target, hostedZone)
        break
      }
      default:
        console.error(`Unexpected Request Type: ${type}`)
        throw new Error('Unexpected Request Type')
    }
    console.log('Route53 successfully updated')
    const responseStatus = 'SUCCESS'
    return sendCfnResponse(lambdaEvent, context, responseStatus, responseData)
  } catch (error) {
    console.error(`Error: ${error}`)
    const responseStatus = 'FAILED'
    const responseData = {}
    return sendCfnResponse(lambdaEvent, context, responseStatus, responseData)
  }
}
