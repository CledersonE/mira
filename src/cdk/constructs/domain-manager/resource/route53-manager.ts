import { AccountPrincipal, CompositePrincipal, Effect, ManagedPolicy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam'
import { Code, Runtime, SingletonFunction } from '@aws-cdk/aws-lambda'
import { SnsEventSource } from '@aws-cdk/aws-lambda-event-sources'
import { Topic } from '@aws-cdk/aws-sns'
import { CfnOutput, Construct, Duration } from '@aws-cdk/core'
import path from 'path'
import { MiraConfig } from '../../../..'
import { MiraStack } from '../../../stack'
import { ROUTE53_MANAGER_POLICY_BOUNDARY } from './route53-manager-access-role'

export const DOMAIN_SUBSCRIPTION_TOPIC = 'DomainSubscriptionTopic'
export const CROSS_ACCOUNT_DOMAIN_MANAGER_ROLE = 'CrossAccountDomainManagerRole'

export class Route53Manager extends MiraStack {
  constructor (parent: Construct) {
    super(parent, 'Route53Manager')

    const { hostedZoneId } = MiraConfig.getDomainConfig()

    if (!hostedZoneId) {
      throw new Error('"hostedZoneId" is required in "domain" config')
    }

    const allowedPrincipals = MiraConfig.getDomainAllowedPrincipals().map(account => new AccountPrincipal(account.env.account))

    // Create role & permissions
    const permissionsBoundary = ManagedPolicy.fromManagedPolicyName(
      this,
      'Route53ManagerPermissionsBoundary',
      MiraConfig.calculateSharedResourceName(ROUTE53_MANAGER_POLICY_BOUNDARY)
    )

    const route53ManagerRole = new Role(this, 'Route53ManagerRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      permissionsBoundary
    })

    route53ManagerRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      resources: [`arn:aws:route53:::hostedzone/${hostedZoneId}`],
      actions: ['route53:ChangeResourceRecordSets']
    }))

    route53ManagerRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: ['route53:GetChange', 'route53:ListResourceRecordSets']
    }))

    route53ManagerRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'))

    // Create topic
    const domainSubscriptionTopic = new Topic(this, DOMAIN_SUBSCRIPTION_TOPIC, {
      displayName: 'Domain Subscription Topic',
      topicName: MiraConfig.getBaseStackName(DOMAIN_SUBSCRIPTION_TOPIC)
    })

    domainSubscriptionTopic.addToResourcePolicy(new PolicyStatement({
      principals: allowedPrincipals,
      effect: Effect.ALLOW,
      resources: [domainSubscriptionTopic.topicArn],
      actions: ['sns:Publish']
    }))

    // Create lambda handler
    const route53ManagerLambda = new SingletonFunction(this, 'Route53ManagerLambda', {
      code: Code.fromAsset(path.join(__dirname, '../lambda')),
      handler: 'route53-manager.handler',
      runtime: Runtime.NODEJS_12_X,
      timeout: Duration.minutes(5),
      uuid: 'ec40983d-c7fe-40fd-a99d-47e61eb307f8',
      environment: {
        HOSTED_ZONE: hostedZoneId
      },
      role: route53ManagerRole
    })

    route53ManagerLambda.addEventSource(new SnsEventSource(domainSubscriptionTopic))

    new CfnOutput(this, 'domainSubscriptionTopicArn', {
      value: domainSubscriptionTopic.topicArn
    })

    // Set up CrossAccountDomainManagerRole
    // this role will be assumed by certificate manager
    const allowedCompositePrincipals = new CompositePrincipal(
      ...MiraConfig
        .getDomainAllowedPrincipals()
        .map(account => new AccountPrincipal(account.env.account))
    )
    const crossAccountDomainManagerRole = new Role(this, CROSS_ACCOUNT_DOMAIN_MANAGER_ROLE, {
      assumedBy: allowedCompositePrincipals,
      roleName: MiraConfig.calculateSharedResourceName(CROSS_ACCOUNT_DOMAIN_MANAGER_ROLE),
      permissionsBoundary
    })
    crossAccountDomainManagerRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      resources: [`arn:aws:route53:::hostedzone/${hostedZoneId}`],
      actions: ['route53:ChangeResourceRecordSets']
    }))
    crossAccountDomainManagerRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: ['route53:GetChange']
    }))
    new CfnOutput(this, 'crossAccountDomainManagerRoleArn', {
      value: crossAccountDomainManagerRole.roleArn
    })
  }
}
