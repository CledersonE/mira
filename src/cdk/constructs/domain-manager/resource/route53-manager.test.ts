import { SynthUtils } from "@aws-cdk/assert"
import { Stack, App } from "@aws-cdk/core"

import {
  MiraConfig,
  DomainConfig,
  Account,
} from "../../../../config/mira-config"

import { Route53Manager } from "./route53-manager"

describe("Route53Manager", () => {
  it("throws when hostedZoneId is not set in config", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("Route53Manager"),
      {}
    )

    MiraConfig.getDomainConfig = (): DomainConfig => ({
      accounts: [],
    })

    MiraConfig.getEnvironment = (): Account => ({
      name: "some-name",
      profile: "some-profile",
      env: { account: "12345", region: "eu-west-1" },
    })

    expect(() => new Route53Manager(stack)).toThrowError(
      '"hostedZoneId" is required in "domain" config'
    )
  })

  test("renders output correctly", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("Route53Manager"),
      {}
    )
    MiraConfig.getDomainConfig = (): DomainConfig => ({
      hostedZoneId: 'Z123456789',
      accounts: [
        'account1',
        'domain'
      ],
    })
    MiraConfig.getDomainAllowedPrincipals = (): any => {
      return [{
        name: "account1",
        profile: "account1",
        env: { account: "account1", region: "eu-west-1" },
      }, {
        name: "domain",
        profile: "domain",
        env: { account: "domain", region: "eu-west-1" },
      }]
    }

    MiraConfig.getEnvironment = (envName: string): Account => {
      if (envName === 'domain') {
        return {
          name: "domain",
          profile: "domain",
          env: { account: "domain", region: "eu-west-1" },
        }
      }

      return {
        name: "account1",
        profile: "account1",
        env: { account: "account1", region: "eu-west-1" },
      }
    }
    const rm = new Route53Manager(stack)
    expect(SynthUtils.toCloudFormation(rm)).toMatchSnapshot()
  })
})
