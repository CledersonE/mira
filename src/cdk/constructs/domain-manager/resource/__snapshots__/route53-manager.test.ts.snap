// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route53Manager renders output correctly 1`] = `
Object {
  "Outputs": Object {
    "crossAccountDomainManagerRoleArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "CrossAccountDomainManagerRole39842D76",
          "Arn",
        ],
      },
    },
    "domainSubscriptionTopicArn": Object {
      "Value": Object {
        "Ref": "DomainSubscriptionTopic75D6B9C5",
      },
    },
  },
  "Parameters": Object {
    "referencetoJohnMyGreatAppRoute53ManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3Bucket32B2FC70Ref": Object {
      "Type": "String",
    },
    "referencetoJohnMyGreatAppRoute53ManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3VersionKeyED2D0E53Ref": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "CrossAccountDomainManagerRole39842D76": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Array [
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::account1:root",
                      ],
                    ],
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::domain:root",
                      ],
                    ],
                  },
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PermissionsBoundary": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":policy/John-MyGreatApp-account1-Route53ManagerPolicyBoundary",
            ],
          ],
        },
        "RoleName": "John-MyGreatApp-account1-CrossAccountDomainManagerRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "CrossAccountDomainManagerRoleDefaultPolicy60DA0FD5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "route53:ChangeResourceRecordSets",
              "Effect": "Allow",
              "Resource": "arn:aws:route53:::hostedzone/Z123456789",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CrossAccountDomainManagerRoleDefaultPolicy60DA0FD5",
        "Roles": Array [
          Object {
            "Ref": "CrossAccountDomainManagerRole39842D76",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DomainSubscriptionTopic75D6B9C5": Object {
      "Properties": Object {
        "DisplayName": "Domain Subscription Topic",
        "TopicName": "John-MyGreatApp-DomainSubscriptionTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "DomainSubscriptionTopicPolicyADFDE884": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Array [
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::account1:root",
                      ],
                    ],
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::domain:root",
                      ],
                    ],
                  },
                ],
              },
              "Resource": Object {
                "Ref": "DomainSubscriptionTopic75D6B9C5",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "DomainSubscriptionTopic75D6B9C5",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "Route53ManagerLambdaDomainSubscriptionTopicB3DF4206": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "SingletonLambdaec40983dc7fe40fda99d47e61eb307f86E8A690C",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "DomainSubscriptionTopic75D6B9C5",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Route53ManagerRole41BB9CC5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "PermissionsBoundary": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":policy/John-MyGreatApp-account1-Route53ManagerPolicyBoundary",
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "Route53ManagerRoleDefaultPolicy7A618D75": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "route53:ChangeResourceRecordSets",
              "Effect": "Allow",
              "Resource": "arn:aws:route53:::hostedzone/Z123456789",
            },
            Object {
              "Action": Array [
                "route53:GetChange",
                "route53:ListResourceRecordSets",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "Route53ManagerRoleDefaultPolicy7A618D75",
        "Roles": Array [
          Object {
            "Ref": "Route53ManagerRole41BB9CC5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SingletonLambdaec40983dc7fe40fda99d47e61eb307f86E8A690C": Object {
      "DependsOn": Array [
        "Route53ManagerRoleDefaultPolicy7A618D75",
        "Route53ManagerRole41BB9CC5",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "referencetoJohnMyGreatAppRoute53ManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3Bucket32B2FC70Ref",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "referencetoJohnMyGreatAppRoute53ManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3VersionKeyED2D0E53Ref",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "referencetoJohnMyGreatAppRoute53ManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3VersionKeyED2D0E53Ref",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "HOSTED_ZONE": "Z123456789",
          },
        },
        "Handler": "route53-manager.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "Route53ManagerRole41BB9CC5",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SingletonLambdaec40983dc7fe40fda99d47e61eb307f8AllowInvokeJohnMyGreatAppRoute53ManagerRoute53Manager1DomainSubscriptionTopic9C2685AE76ADEA39": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "SingletonLambdaec40983dc7fe40fda99d47e61eb307f86E8A690C",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "DomainSubscriptionTopic75D6B9C5",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
